### DISCUSSION TITLE
Here's a discussion of rust-gb!

### DISCUSSION THREAD

**Discussion by zlfn** [zlfn/rust-gb#1] [Announcements]
Feel free to write here if you are interested in this project and want to be involved. (Any writing is acceptable!)

Minor questions are good, improvements are good, and significant discussions are fine.

**Comment by nixbytes** [comment]
I think it would be a good idea to have very basic examples of games. classic like pong and alien shooter to get started

**Reply by zlfn** [reply]
I agree. But there's no API for VRAM in this project yet, so it's not possible right now. It's on the next to-do list, so I think it'll be possible in the near future.

**Reply by nixbytes** [reply]
totally understand, are their plans to C API from VRAM with rust as a temporary solution?

**Reply by nixbytes** [reply]
we might need to have rust write the assembly. I can try to look into this, no promises :)

**Reply by zlfn** [reply]
> totally understand, are their plans to C API from VRAM with rust as a temporary solution?

The C API for VRAM is easy to use right now. There is no mapping yet, but this library is designed to use all of GBDK's functions, so we simply have to map some extern functions.

**Reply by nixbytes** [reply]
I will need to read up on these things and try it out when time permits

**Comment by K4stor** [comment]
Hi I want to learn rust and gameboy development, so gb rust could be a perfect start !

I just wonder, having seen the people who use C for gb development, still write some functions in assembly for maximum performance, 
is that possible too with rust ?

Like having inline assembly ?

**Reply by zlfn** [reply]
Currently, Rust-GB is almost a Proof of Concept (PoC). The ultimate goal is to create a toolkit that allows developing Game Boy games in Rust, but at this stage, it is far from stable and remains highly experimental. (Of course, contributions are very welcome!)

As you can see in the `README.md`, Rust-GB uses a complex and somewhat hacky toolchain to build Rust code for the Game Boy.
Because the **Z80-family CPU is not an officially supported target of Rust/LLVM**, **inline assembly is not possible**.
However, during the build process, Rust-GB **links external assembly functions from the `/asm` folder and C functions from the `/c` folder** into the final ROM file.
This means inline assembly cannot be used, but you can still call assembly functions at the function level.
(While this introduces some overhead from function calls and stack operations, it is sufficient to replace inefficiently compiled instructions.)

Currently, external assembly code is used for certain special instructions — such as `halt`, `ei`, and `di` — that cannot be compiled via LLVM-CBE.
For details, see: [https://github.com/zlfn/rust-gb/tree/main/ext/asm](https://github.com/zlfn/rust-gb/tree/main/ext/asm).
